src/common/imageio_webp.c

```diff
@@ -0,0 +1,121 @@
+/*
+    This file is part of darktable,
+    Copyright (C) 2022 darktable developers.
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <inttypes.h>
+
+#include <webp/decode.h>
+#include <webp/mux.h>
+
+#include "image.h"
+#include "imageio.h"
+
+dt_imageio_retval_t dt_imageio_open_webp(dt_image_t *img, const char *filename, dt_mipmap_buffer_t *mbuf)
+{
+  int w, h;
+  WebPMux *mux;
+  WebPData wp_data;
+  WebPData icc_profile;
+
+  FILE *f = g_fopen(filename, "rb");
+  if(!f)
+  {
+    fprintf(stderr,"[webp_open] cannot open file for read: %s\n", filename);
+    return DT_IMAGEIO_FILE_CORRUPTED;
+  }
+
+  fseek(f, 0, SEEK_END);
+  size_t filesize = ftell(f);
+  fseek(f, 0, SEEK_SET);
+
+  void *read_buffer = g_malloc(filesize);
+
+  if(fread(read_buffer, 1, filesize, f) != filesize)
+  {
+    fclose(f);
+    g_free(read_buffer);
+    fprintf(stderr,"[webp_open] failed to read %zu bytes from %s\n", filesize, filename);
+    return DT_IMAGEIO_FILE_CORRUPTED;
+  }
+  fclose(f);
+
+// WebPGetInfo will tell us image dimensions needed for dt image buffer allocation
+  if(!WebPGetInfo(read_buffer, filesize, &w, &h))
+  {
+    g_free(read_buffer);
+    fprintf(stderr,"[webp_open] WebPGetInfo failed\n");
+    return DT_IMAGEIO_FILE_CORRUPTED;
+  }
+  img->width = w;
+  img->height = h;
+  img->buf_dsc.channels = 4;
+  img->buf_dsc.datatype = TYPE_FLOAT;
+
+  float *mipbuf = (float *)dt_mipmap_cache_alloc(mbuf, img);
+  if(!mipbuf)
+  {
+    g_free(read_buffer);
+    fprintf(stderr, "[webp_open] could not alloc full buffer for image: %s\n", img->filename);
+    return DT_IMAGEIO_CACHE_FULL;
+  }
+
+  uint8_t *int_RGBA_buf = WebPDecodeRGBA(read_buffer, filesize, &w, &h);
+  if(!int_RGBA_buf)
+  {
+    g_free(read_buffer);
+    fprintf(stderr,"[webp_open] failed to decode file: %s\n", filename);
+    return DT_IMAGEIO_FILE_CORRUPTED;
+  }
+
+  uint8_t intval;
+  float floatval;
+
+  for(int i=0; i < w*h*4; i++)
+  {
+    intval = *(int_RGBA_buf+i);
+    floatval = intval / 255.f;
+    *(mipbuf+i) = floatval;
```

Comment:
And if there was? Although dt should ignore alpha in principle, I'm not sure all the code paths have been that robustly tested...

Reply:
> In the source code of the webp library, you can see that the buffer is filled with zeros (using memset). Therefore, if there is no alpha channel in the image, it will be zero in the decoded data buffer. There is no need to write zeros separately.

Well if this is changed at some point? I mean the source is not the documentation. So if in lib WebP it is said that the alpha is set to 0 then ok.

Reply:
I was replying too quickly to point out that the library doesn't leave this channel uninitialized (at least that's how I understood kmilos' concern about zeroing this channel). But I, without thinking, mistakenly wrote that in normal (opaque) images this channel will remain zero. Just the opposite. This is an alpha channel, so by definition it represents the degree of _opacity_. In other words, in normal (without transparency) images, this channel will contain the maximum values ​​(in our case, 255), and the full transparency of the pixel corresponds to the value 0. Now I checked the state of the buffer in the debugger after reading normal images and images with transparency - this is indeed the case.

After implementing the import of the webp images, I tested processing of such image in dt, in parallel with the processing of the jpeg image that was its source. In my tests everything works fine and makes no difference with jpeg processing, although the 4th channel after reading the webp contained (after converting to float) 1 instead of 0. After kmilos' comment, I also added 4th channel zeroing and see no difference in working with such data. The question arises, does the content of this channel affect anything at all?

> Although dt should ignore alpha in principle, I'm not sure all the code paths have been that robustly tested...

If the data in the 4th channel is really ignored, as I assumed and you confirm, then what tests of execution paths can we be talking about? We just don't use the data from this channel anywhere in the image processing, that's all. As I see it, channel 4 is only needed for alignment, which allows speedup due to SSE operations. No?

Don't get me wrong, there's no problem adding this zeroing, although it slows down the image reading process, even if not by much, I just don't see what purpose we're slowing down for.

Reply:
AFAIK the 4th channel is used in masks, e.g. when displaying one in darkroom.